### PR TITLE
fix: ackee audit

### PR DESCRIPTION
**Pull request by mfw78**
This is a WIP PR. Do not merge yet.

This PR:

1. Applies fixes, with commits in the PR tagged with the ackee audit finding that they fix.
2. Applies additional fixes around consistency (comments, a 'utils' library for common functions in conditional orders).
3. Adds in `appData` to all conditional order types for consistency.
4. Generalises the deployer so that the `salt` for `CREATE2` is configurable via environment variable.

**Todo**:

- [ ]  ~~Oracle data checking in `StopLoss`~~ Tracked in #30

### CODE REVIEW

File: script/deploy_ProdStack.s.sol [hunk 1]
```diff
@@ -19,19 +19,20 @@ contract DeployProdStack is Script {
     function run() external {
         uint256 deployerPrivateKey = vm.envUint("PRIVATE_KEY");
         address settlement = vm.envAddress("SETTLEMENT");
+        bytes32 salt = vm.envBytes32("SALT");
 
         vm.startBroadcast(deployerPrivateKey);
 
         // Deploy ExtensibleFallbackHandler
-        new ExtensibleFallbackHandler{salt: "v1"}();
+        new ExtensibleFallbackHandler{salt: salt}();
```
Comment (fleupold) [inline script/deploy_ProdStack.s.sol:27]: By having the salt hardcoded and checked into code, we can reproduce the builds if needed later (e.g. on other chains). If we pass it as an env, do we need to think about a way to document the used salt (or can it be easily recovered from the deployment tx)?

File: src/types/StopLoss.sol [hunk 2]
```diff
@@ -44,33 +44,31 @@ contract StopLoss is BaseConditionalOrder {
...
+    function getTradeableOrder(address, address, bytes32, bytes calldata staticInput, bytes calldata)
+        public
+        view
+        override
+        returns (GPv2Order.Data memory order)
+    {
         Data memory data = abi.decode(staticInput, (Data));
-        (, int256 latestSellPrice, , , ) = data.sellTokenPriceOracle.latestRoundData();
-        (, int256 latestBuyPrice, , , ) = data.buyTokenPriceOracle.latestRoundData();
+        (, int256 latestSellPrice,,,) = data.sellTokenPriceOracle.latestRoundData();
```
Comment (fleupold) [inline src/types/StopLoss.sol:54]: Make sense to do this in a separate PR as it is more involved, but I assume they suggest to use check that the oracle's decimals view method returns the same number for both and check that updated at is not too far in the past (not sure if we should allow the user to specify this though)

File: test/ComposableCoW.base.t.sol [hunk 3]
```diff
@@ -149,10 +149,16 @@ contract BaseComposableCoWTest is Base, Merkle {
...
+        vm.startPrank(owner);
         bytes32 orderHash = keccak256(abi.encode(params));
+        bytes32 ctx = composableCow.cabinet(owner, orderHash);
         composableCow.remove(orderHash);
         assertEq(composableCow.singleOrders(owner, orderHash), false);
+        if (ctx != bytes32(0)) {
+            // ensure that the context was cleared
+            assertEq(composableCow.cabinet(owner, orderHash), bytes32(0));
+        }
+        vm.stopPrank();
```
Comment (fleupold) [inline test/ComposableCoW.base.t.sol:161]: Why do we need to start/stop prank?

File: test/ComposableCoW.twap.t.sol [hunk 4]
```diff
@@ -322,6 +322,10 @@ contract ComposableCoWTwapTest is BaseComposableCoWTest {
                 GPv2Order.hash(part, settlement.domainSeparator()), signature
             ) == ERC1271.isValidSignature.selector
         );
+
+        // Now remove the order
+        vm.prank(address(safe1));
+        composableCow.remove(keccak256(abi.encode(params)));
```
Comment (fleupold) [inline test/ComposableCoW.twap.t.sol:328]: do we want to assert anything afterwards or are we simply fine with it not reverting (seems weird to test an action without checking a post-assumption)?

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia0.giphy.com%2Fmedia%2FdaxVu1mZErt2qKYdtX%2Fgiphy.gif&token=fix" data-gitmeme-token="fix"></a>

File: test/ComposableCoW.base.t.sol (hunk 3 above)
Comment (mfw78) [reply test/ComposableCoW.base.t.sol:161]: Ahh, this is because the `composableCow.cabinet` call took the `vm.prank()`. Instead I'll move the `vm.prank()` down to precede `composableCoW.remove` and remove the `vm.stopPrank()`

File: src/types/StopLoss.sol (hunk 2 above)
Comment (mfw78) [reply src/types/StopLoss.sol:54]: Yes, I'll put this into a separate PR. tl;dr I'm working on actually generalising the strike price such that it's sell token / buy token in ERC20 decimals, and normalising the returned quote from the oracles.

File: script/deploy_ProdStack.s.sol (hunk 1 above)
Comment (mfw78) [reply script/deploy_ProdStack.s.sol:27]: Fair point. I'll revert this one. I was hoping that it was easy to recover from the broadcast JSON but I can't see it easily.

File: test/ComposableCoW.twap.t.sol (hunk 4 above)
Comment (mfw78) [reply test/ComposableCoW.twap.t.sol]: I think from memory there were assertions in another test suite referencing this, but in lieu of that, I've added assertions afterwards to stick to conventions.

**Review by fleupold — APPROVED** [review]
Looks great!
